core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServices.java

```diff
@@ -307,25 +318,69 @@ private static ResponseConverter converter(Method method) {
+        final Converter[] converters = clazz.getAnnotationsByType(Converter.class);
+        final ImmutableMap.Builder<Class<?>, ResponseConverter> builder = ImmutableMap.builder();
         for (Converter converter : converters) {
-            Class<?> target = converter.target();
+            final Class<?> target = converter.target();
             if (target == Unspecified.class) {
                 throw new IllegalArgumentException(
                         "@Converter annotation must have a target type specified.");
             }
-            try {
-                ResponseConverter instance = converter.value().newInstance();
-                builder.put(target, instance);
-            } catch (Exception e) {
-                throw new IllegalStateException(e);
-            }
+            builder.put(target, newInstance(converter.value()));
         }
         return builder.build();
     }
 
     /**
+     * Returns a decorator chain which is specified by {@link Decorate} annotations.
+     */
+    private static Function<Service<HttpRequest, HttpResponse>,
+            ? extends Service<HttpRequest, HttpResponse>> decorator(Method method) {
+
+        final Decorate decorate = method.getAnnotation(Decorate.class);
+        if (decorate == null) {
+            return Function.identity();
+        }
+
+        final Class<? extends DecoratingServiceFunction<HttpRequest, HttpResponse>>[] c = decorate.value();
+        if (c.length == 0) {
+            return Function.identity();
+        }
+
+        // Respect the order of decorators which is specified by a user. The first one is first applied.
+        Function<Service<HttpRequest, HttpResponse>,
+                ? extends Service<HttpRequest, HttpResponse>> decorator = newDecorator(c[c.length - 1]);
+        for (int i = c.length - 2; i >= 0; i--) {
+            decorator = decorator.andThen(newDecorator(c[i]));
+        }
+        return decorator;
+    }
+
+    /**
+     * Returns a new decorator which decorates a {@link Service} by {@link FunctionalDecoratingService}
+     * and the specified {@link DecoratingServiceFunction}.
+     */
+    private static Function<Service<HttpRequest, HttpResponse>,
+            ? extends Service<HttpRequest, HttpResponse>> newDecorator(
+            Class<? extends DecoratingServiceFunction<HttpRequest, HttpResponse>> clazz) {
+        return service -> new FunctionalDecoratingService<>(
+                service, decoratingServiceFunctions.computeIfAbsent(clazz, AnnotatedHttpServices::newInstance));
+    }
+
+    /**
+     * Returns a new instance of the specified {@link Class}.
+     */
+    private static <T> T newInstance(Class<? extends T> clazz) {
+        try {
+            final Constructor<? extends T> constructor = clazz.getDeclaredConstructor();
+            constructor.setAccessible(true);
+            return constructor.newInstance();
+        } catch (Exception e) {
+            throw new IllegalStateException(e);
```

Comment:
How about adding a message to the `IllegalStateException`, e.g.

    throw new IllegalStateException("A decorator function class specified in @" + 
                                    Decorate.class.getSimpleName() + 
                                    " annotation must have an accessible default constructor:" +
                                    clazz.getName(), e);

..?